pkg/controller/beat/common/config.go

```diff
@@ -89,34 +95,81 @@ func buildBeatConfig(
 		return nil, err
 	}
 
-	userConfig := beat.Spec.Config
+	// get user config from `config` or `configRef`
+	userConfig, err := getUserConfig(params)
+	if err != nil {
+		return nil, err
+	}
+
 	if userConfig == nil {
 		return cfg.Render()
 	}
 
-	userCfg, err := settings.NewCanonicalConfigFrom(userConfig.Data)
-	if err != nil {
+	if err = cfg.MergeWith(userConfig); err != nil {
+		return nil, err
+	}
+
+	return cfg.Render()
+}
+
+// getUserConfig extracts the config either from the spec `config` field or from the Secret referenced by spec
+// `configRef` field.
+func getUserConfig(params DriverParams) (*settings.CanonicalConfig, error) {
+	if params.Beat.Spec.Config != nil {
+		userCfg, err := settings.NewCanonicalConfigFrom(params.Beat.Spec.Config.Data)
+		if err != nil {
+			return nil, err
+		}
+		return userCfg, nil
```

Comment:
I think we can `return settings.NewCanonicalConfigFrom(params.Beat.Spec.Config.Data)` directly.